Updating --include-path-patterns and --exclude-path-patterns to use a table of tables

mayuriesha opened the pull request:
To help us get this pull request reviewed and merged quickly, please be sure to include the following items:

* [x] Tests (if applicable)
* [x] Documentation (if applicable)
* [x] Changelog entry
* [x] A full explanation here in the PR description of the work done

## PR Type
What kind of change does this PR introduce?

* [ ] Bugfix
* [x] Feature
* [ ] Code style update (formatting, local variables)
* [ ] Refactoring (no functional changes, no api changes)
* [ ] Build related changes
* [ ] CI related changes
* [ ] Documentation content changes
* [ ] Tests
* [ ] Other

## Backward Compatibility

Is this change backward compatible with the most recently released version? Does it introduce changes which might change the user experience in any way? Does it alter the API in any way?

* [x] Yes (backward compatible)
* [ ] No (breaking changes)


## Issue Linking
closes #257 

## What's new?
- Supports table of tables format for `--include-path-patterns` and `--exclude-path-patterns` with an option to specify the reason.
Ex: include-path-patterns = [{path-pattern="path-pattern", reason="reason to include path patterns"}]
- Warns users about deprecation of old format
- Making sure either one the formats is used until the old format is completely deprecated

mayuriesha committed:
Supporting table of contents for include and exclude path patterns

mayuriesha committed:
Updating unit tests and documentation

mayuriesha committed:
Updating logic for inlcude and exclude paths to use two different formats

mayuriesha committed:
Adding try catch

mayuriesha committed:
Minor wording changes

mayuriesha committed:
Merge remote-tracking branch 'upstream/v3.x' into path_patterns

mayuriesha committed:
Updating Changelog

tarkatronic commented on tartufo/scanner.py:
```diff
@@ -247,13 +248,38 @@ def included_paths(self) -> List[Pattern]:
+                    "Old format of --include-path-patterns option and config file setup include-path-patterns "
+                    "= ['inclusion pattern'] has been deprecated and will be removed in a future version. "
+                    "Make sure all the inclusions are set up using new pattern i.e. include-path-patterns = "
+                    "[path-pattern='inclusion pattern',reason='reason for inclusion'] in the config file",
```
```suggestion
                    "[{path-pattern='inclusion pattern',reason='reason for inclusion'},] in the config file",
```

tarkatronic commented on tartufo/scanner.py:
```diff
@@ -247,13 +248,38 @@ def included_paths(self) -> List[Pattern]:
+                    if patterns
+                    else []
+                )
+                return self._included_paths
```
Not that it's really a problem, but this and line 277 really shouldn't be necessary. They're a bit redundant, since line 283 always returns this same value.

mayuriesha replied:
I believe we need this return as we need to stop further execution if the user has new style of include and exclude path patterns

rbailey-godaddy replied:
So, a couple of things. We're stuck with butt-ugly code because of the logic, but I think a couple of tactical maneuvers could help us out. These comments are inclusive for the block from L259 through L298.

First, we have a problem in that although we are telling people they must use "all old flavor" or "all new flavor", the code actually enforces "all processed toml files must use the same old or new flavor" which isn't what we say, may not be practicable, and isn't necessary.

To combat this, I think we need to process the `patterns` (global) and `include_patterns` (local) lists independently, and combine them at the end after they have been compiled. Tactically, I would also suggest avoiding reuse of `patterns` when we know the type is changing and use of `set` rather than `list` for the result so we don't have to deduplicate manually (and subsequent queries should be more efficient).

In pseudocode, something more like:

```
self._included_paths: Set[Pattern] = set()
for plist in (self.global_options.include_path_patterns, self.config_data.get("include_path_patterns")):
    if plist:
        try:
            self._included_paths.update(
                config.compile_path_rules({x["path-patterns"] for x in plist})
            )
            continue # New way worked, move on
        except TypeError:
            warnings.warn(blah blah blah, DeprecationWarning)
        # Try the old way
        try:
            self._included_paths.update(
                config.compile_path_rules(set(plist))
            )
        except TypeError as exc:
             raise types.ConfigException(you can't mix both types in the same list)
```
We possibly do some extra regex compiling before deduplicating, but I think the clarity is worth the overhead.

rbailey-godaddy replied:
...and I'd use the same pattern (excuse the pun) for the exclude processing that comes next.

rbailey-godaddy replied:
...and yes, this changes the return from `List[Pattern]` to `Set[Pattern]` but I can't see anything in the code base cares, and I would expect set membership testing to be more efficient than list membership testing.

tarkatronic commented on tartufo/scanner.py:
```diff
@@ -247,13 +248,38 @@ def included_paths(self) -> List[Pattern]:
+                raise types.ConfigException(
+                    f"Combination of old and new format of include-path-patterns will not be supported.\n{exc}"
+                )
+        self.logger.debug("Included paths was initialized as: %s", self._included_paths)
```
```suggestion
            self.logger.debug("Included paths was initialized as: %s", self._included_paths)
```
We don't need this to log every single time `included_paths` is accessed 😄

tarkatronic commented on tartufo/scanner.py:
```diff
@@ -280,12 +306,37 @@ def excluded_paths(self) -> List[Pattern]:
+                raise types.ConfigException(
+                    f"Combination of old and new format of exclude-path-patterns will not be supported.\n{exc}"
+                )
+        self.logger.debug("Excluded paths was initialized as: %s", self._excluded_paths)
```
All the same comments apply here as in `included_paths` above.

tarkatronic commented on tartufo/scanner.py:
```diff
@@ -692,27 +743,6 @@ def load_repo(self, repo_path: str) -> pygit2.Repository:
             config_file = None
         if config_file and config_file != self.global_options.config:
             self.config_data = data
-            include_patterns = list(data.get("include_path_patterns", ()))
-            repo_include_file = data.get("include_paths", None)
-            if repo_include_file:
-                repo_include_file = pathlib.Path(repo_path, repo_include_file)
-                if repo_include_file.exists():
-                    with repo_include_file.open() as handle:
-                        include_patterns += handle.readlines()
-            if include_patterns:
-                include_patterns = config.compile_path_rules(include_patterns)
-                self._included_paths = list(set(self.included_paths + include_patterns))
-
-            exclude_patterns = list(data.get("exclude_path_patterns", ()))
-            repo_exclude_file = data.get("exclude_paths", None)
-            if repo_exclude_file:
-                repo_exclude_file = pathlib.Path(repo_path, repo_exclude_file)
-                if repo_exclude_file.exists():
-                    with repo_exclude_file.open() as handle:
-                        exclude_patterns += handle.readlines()
-            if exclude_patterns:
-                exclude_patterns = config.compile_path_rules(exclude_patterns)
-                self._excluded_paths = list(set(self.excluded_paths + exclude_patterns))
```
With this code block gone, how do these supplemental patterns now get added to the config?

rbailey-godaddy replied:
I think L760 is handling this, although I'm fuzzy on why the config_data property and setter were removed just above (or alternatively why they existed previously).

emayuri-godaddy replied:
There is another config_data property at line 244. So I am removing the duplicate one here.

tarkatronic commented on tests/test_git_repo_scanner.py:
```diff
@@ -61,24 +61,26 @@ def test_load_repo_does_not_filter_submodules_when_requested(
         self.assertCountEqual(
             test_scanner.included_paths,
-            [re.compile("foo/"), re.compile("tartufo/"), re.compile("scripts/")],
+            [re.compile("foo/")],
```
This test doesn't seem to me like it is proving the same intent anymore. Especially now that you added the `self.global_options.include_path_patterns = ("foo/",)`

Now, this test only shows that we load patterns from global options. The intent of this test is to show that:

* `global_options` include `"foo/"`
* `load_repo` is called with a separate repository, with a config file in it.
* That config file includes `"bar/"`
* The `included_paths` must now contain both `"foo/"`, and `"bar/"`

tarkatronic commented on tests/test_git_repo_scanner.py:
```diff
@@ -61,24 +61,26 @@ def test_load_repo_does_not_filter_submodules_when_requested(
-            {"exclude_paths": "exclude-files", "exclude_path_patterns": ("bar/",)},
+            {"exclude_path_patterns": ("bar/",)},
         )
+        self.global_options.exclude_path_patterns = ("bar/",)
```
Same comment as the test above.

mayuriesha committed:
Addressing review comments on deprecation message

mayuriesha committed:
Processing the patterns added in the config as well

mayuriesha committed:
Updating unit tests to verify patterns are processed from config and global options

mayuriesha committed:
Updating include-path-patterns

rbailey-godaddy commented on tartufo/cli.py:
```diff
@@ -92,19 +92,23 @@ def get_command(self, ctx: click.Context, cmd_name: str) -> Optional[click.Comma
-    include in the scan. This option can be specified multiple times to include
-    multiple patterns. If not provided (default), all Git object paths are
-    included unless otherwise excluded via the --exclude-path-patterns
+    include in the scan. Multiple patterns can be included in the config file using
+    include-path-patterns = [{path-pattern="pattern", reason="reason to include pattern},].
+    If not provided (default), all Git object paths
+    are included unless otherwise excluded via the --exclude-path-patterns
```
I'm picking arbitrarily on this one but also consider L109-L111. Can at least one of these mention what happens when a file matches both `include-path-patterns` and `exclude-path-patterns`, i.e. which takes precedence?

rbailey-godaddy replied:
I see while researching other changes below that a file which is both included and excluded will be excluded.

rbailey-godaddy reviewed (changes requested):
Pretty much one change to relax unintended restriction that needlessly crosses multiple configuration files.

mayuriesha commented:
@rbailey-godaddy I have created an issue to address the last few review comments as part of another PR after 3.0 is released

rbailey-godaddy reviewed (approved):
LGTM (keeping in mind the new backlog story)

rscottbailey committed:
Update path processing

Clean up `included_paths` and `excluded_paths` with basically the same
logic used for #290.

rbailey-godaddy committed:
Merge branch 'v3.x' into path_patterns

rscottbailey committed:
reformat merged test file with black

tarkatronic reviewed (approved):
🎉
